### PR TITLE
Reworking timing changes to be more flexible.

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu-framework/pull/777
- [x] https://github.com/ppy/osu-framework/pull/788
- [x] https://github.com/ppy/osu-framework/pull/810

I made this generic enough so that it (hopefully) should be usable in taiko with no major issues (supports both y-relative and x-relative movement).

Does not currently support:
- Adding new timing changes (after all the hit objects have been added).
- Removing timing changes.
- Reordering of timing changes.
- Reordering of hit objects (between multiple timing changes).

### CODE REVIEW

File: osu.Desktop.VisualTests/Tests/TestCaseManiaPlayfield.cs [hunk 1]
```diff
@@ -39,70 +42,83 @@ public override void Reset()
...
+
+                return new DrawableScrollingTimingChange(new TimingChange
+                {
+                    BeatLength = 1000,
+                    Time = time
+                });
             };
 
-            Action createPlayfieldWithNotesAcceptingInput = () =>
+            Action<bool> createPlayfieldWithNotes = g =>
```
Comment (peppy) [inline osu.Desktop.VisualTests/Tests/TestCaseManiaPlayfield.cs]: `g` -> `gravity`?

File: osu.Game.Rulesets.Mania/UI/ManiaHitRenderer.cs [hunk 2]
```diff
@@ -19,25 +20,42 @@
...
     {
-        public int? Columns;
+        private int? columns;
+        public int Columns
+        {
+            get { return columns ?? (int)Math.Round(Beatmap.BeatmapInfo.Difficulty.CircleSize); }
+            set { columns = value; }
+        }
+
+        public Dictionary<int, List<DrawableTimingChange>> HitObjectTimingChanges;
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/UI/ManiaHitRenderer.cs]: there's nothing explaining what the `int` here is

File: osu.Game.Rulesets.Mania/Timing/Drawables/DrawableScrollingTimingChange.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2007-2017 ppy Pty Ltd <contact@ppy.sh>.
+// Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
+
+using osu.Game.Rulesets.Timing;
+
+namespace osu.Game.Rulesets.Mania.Timing.Drawables
+{
+    public class DrawableScrollingTimingChange : DrawableManiaTimingChange
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/Timing/Drawables/DrawableScrollingTimingChange.cs]: `DrawableScrollingManiaTimingChange` maybe. Going based on filename, it's hard to tell it's mania-specific (and the order of inheritance).

Comment (peppy) [inline osu.Game.Rulesets.Mania/Timing/Drawables/DrawableScrollingTimingChange.cs]: also one-line xmldoc wouldn't go astray (just describing what the "scroll" part of this is actually achieving)

File: osu.Game/Rulesets/Timing/TimingChangeContainer.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+using osu.Framework.Graphics.Containers;
+using osu.Game.Rulesets.Objects.Drawables;
+using osu.Game.Rulesets.Timing.Drawables;
+using OpenTK;
+
+namespace osu.Game.Rulesets.Timing
+{
+    public class TimingChangeContainer : Container<DrawableTimingChange>, IHasTimeSpan
+    {
+        public Vector2 TimeSpan { get; set; }
```
Comment (peppy) [inline osu.Game/Rulesets/Timing/TimingChangeContainer.cs]: what is this describing? if it's the length, why does it have two dimensions?

File: osu.Game.Rulesets.Mania/Timing/Drawables/DrawableScrollingTimingChange.cs (hunk 3 above)
Comment (smoogipooo) [inline osu.Game.Rulesets.Mania/Timing/Drawables/DrawableScrollingTimingChange.cs]: But... It's in osu.Game.Rulesets.Mania? I thought we weren't going to use the ruleset in the name unless they conflicted with something in osu.Game?

File: osu.Game.Rulesets.Mania/UI/Column.cs [hunk 5]
```diff
@@ -187,10 +185,17 @@ public Color4 AccentColour
             }
         }
 
-        public void Add(DrawableHitObject<ManiaHitObject, ManiaJudgement> hitObject)
+        public double TimeSpan
```
Comment (peppy) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: sure you don't wanna call this Duration or Length?

Comment (peppy) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: a TimeSpan to me is a .net class which represents a span of time

Comment (smoogipooo) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: Duration sounds like a bit of a misnomer... My thinking was that it's a span of time that you see on the screen.

Duration or Length to me implies that it takes some amount of time to do something, but this just represents how "much" time you see on the screen.

Comment (Tom94) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: Why not be consistent with standard and call it preempt?

Comment (smoogipooo) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: PreEmpt works with me.

Comment (peppy) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: Isn't it the duration of the timing section though?

Comment (smoogipooo) [inline osu.Game.Rulesets.Mania/UI/Column.cs]: No, it's the length of time which the coordinates of the container spans. Saying it that way I guess length might also be okay but PreEmpt is more descriptive imo.

**Comment by smoogipooo** [comment]
Just gonna close this for a little while while I finish up my rework/refactorings.

**Pull request closed without merging**